[lldb][Progress] Separate title and details

chelcassanova opened the pull request:
Per this RFC:
https://discourse.llvm.org/t/rfc-improve-lldb-progress-reporting/75717 on improving progress reports, this commit separates the title field and details field so that the title specifies the category that the progress report falls under. The details field is added as a part of the constructor for progress reports and by default is an empty string. Also updates the test to check for details being correctly reported from the event structured data dictionary.

chelcassanova committed:
[lldb][Progress] Separate title and details

Per this RFC:
https://discourse.llvm.org/t/rfc-improve-lldb-progress-reporting/75717
on improving progress reports, this commit separates the title field and
details field so that the title specifies the category that the progress
report falls under. The details field is added as a part of the
constructor for progress reports and by default is an empty string. Also
updates the test to check for details being correctly reported from the
event structured data dictionary.

chelcassanova committed:
fixup! [lldb][Progress] Separate title and details

Formatting

clayborg commented on lldb/include/lldb/Core/Progress.h:
```diff
@@ -69,7 +69,8 @@ class Progress {
   /// progress is to be reported only to specific debuggers.
-  Progress(std::string title, uint64_t total = UINT64_MAX,
+  Progress(std::string title, std::string details = {},
+           uint64_t total = UINT64_MAX,
```
Might be good to start using `std::optional<uint64_t>` internally in the `Progress` class and avoid the magic numbers as long as we are changing things up

chelcassanova replied:
For sure, this was actually a part of the original patch I made trying to introduce the aggregate flag.

clayborg commented on lldb/source/Plugins/SymbolFile/DWARF/SymbolFileDWARF.cpp:
```diff
@@ -519,8 +519,7 @@ void SymbolFileDWARF::InitializeObject() {
         apple_types.GetByteSize() > 0 || apple_objc.GetByteSize() > 0) {
-      Progress progress(llvm::formatv("Loading Apple DWARF index for {0}",
-                                      module_desc.GetData()));
+      Progress progress("Loading Apple DWARF index", module_desc.GetData());
```
Should we get rid of this? This will never take any time to load and will just cause progress spam. I would vote to remove this unless there is something that can take time, but I can't think of anything.

chelcassanova replied:
This can be removed, I also have to update the progress report for ELF files to match the rest of the changes here 😅

clayborg commented on lldb/source/Plugins/SymbolFile/DWARF/SymbolFileDWARF.cpp:
```diff
@@ -532,8 +531,7 @@ void SymbolFileDWARF::InitializeObject() {
     if (debug_names.GetByteSize() > 0) {
-      Progress progress(
-          llvm::formatv("Loading DWARF5 index for {0}", module_desc.GetData()));
+      Progress progress("Loading DWARF5 index", module_desc.GetData());
```
DWARF5 on the other hand usually has a single .debug_names table per compile unit, and there can be mixed DWARF5 .debug_names tables along with some compile units that need to be manually indexed, so it might be worth keeping this progress.

clayborg reviewed (approved):
The roots of this look really good. A few nits we can choose to fix or do in another patch.

chelcassanova committed:
Update ELF progress report, remove Apple DWARF report, use std::optional

Updates the ELF index progress report, remove the Apple DWARF progress
report and use a std::optional for the `m_total` field in `Progress`

clayborg commented on lldb/source/Core/Progress.cpp:
```diff
@@ -30,18 +34,19 @@ Progress::~Progress() {
+      m_completed = m_total.value();
     else
       m_completed += amount;
     ReportProgress(update);
```
We are now storing any new detail into `m_details`, do we need to send this to ReportProgress now? Prior to this, this string wasn't stored anywhere but now we have a copy, so we should probably remove the parameter to this.

clayborg commented on lldb/source/Core/Progress.cpp:
```diff
@@ -51,9 +56,10 @@ void Progress::Increment(uint64_t amount, std::string update) {
 void Progress::ReportProgress(std::string update) {
```
remove `std::string update` as we have it in `m_details`

clayborg reviewed (commented):
A bit more cleanup, see inline comment

clayborg commented on lldb/source/Core/Progress.cpp:
```diff
@@ -51,9 +56,10 @@ void Progress::Increment(uint64_t amount, std::string update) {
+    // complete, and only modify m_complete is m_total isn't null.
+    if (m_total.has_value())
+      m_complete = m_completed == m_total.value();
+    Debugger::ReportProgress(m_id, m_title, m_details, m_completed,
+                             m_total.value_or(0), m_debugger_id);
```
This will actually mess up reporting of progress reports with no valid total. We will always sent out `completed = 0` and `total = 0` for both the start and end events. The easiest way to fix this would be to leave `m_total` as a uint64_t inside of progress, but still allow the argument to be an optional in the constructor. We need `completed` to not be equal to `total`, the start progress event will have `completed=0` and `total=<non-zero>`, and then the end event will need to have `completed=total` where total is non zero.

chelcassanova replied:
Not sure how I didn't realize that, I had this implementation from my first patch for a bool flag for this. Weirdly enough the test passed with those changes even though I don't think they should've. Either way I'll update this, my mistake.

clayborg commented on lldb/include/lldb/Core/Progress.h:
```diff
@@ -96,13 +97,14 @@ class Progress {
-  /// Total amount of work, UINT64_MAX for non deterministic progress.
-  const uint64_t m_total;
+  /// Total amount of work, use a std::nullopt for non deterministic progress.
+  const std::optional<uint64_t> m_total;
```
We need to leave this as a `uint64_t` so that we can report progress start events and end events correctly. See comments in `Progress::ReportProgress(...)`. We need `total` to be reported as non-zero as the end events are considered a notification where `completed == total`

clayborg commented on lldb/source/Core/Progress.cpp:
```diff
@@ -11,15 +11,19 @@
                    lldb_private::Debugger *debugger)
-    : m_title(title), m_id(++g_id), m_completed(0), m_total(total) {
-  assert(total > 0);
+    : m_title(title), m_details(details), m_id(++g_id), m_completed(0),
+      m_total(total) {
+  assert(total == std::nullopt || total > 0);
```
We need total to be non zero when there is no actual total:
```
m_total(1) { // Default to one if in case there is no valid total
// Set the actual total if we have a valid value
if (total)
  m_total = *total;
```

clayborg commented on lldb/source/Core/Progress.cpp:
```diff
@@ -30,18 +34,19 @@ Progress::~Progress() {
   std::lock_guard<std::mutex> guard(m_mutex);
-  if (!m_completed)
-    m_completed = m_total;
+  if (m_total.has_value() && !m_completed)
+    m_completed = m_total.value();
```
revert (see comments above and down in `Progress::ReportProgress(...)` for details

chelcassanova replied:
Seconded from your comments on `ReportProgress` above.

chelcassanova commented on lldb/source/Core/Progress.cpp:
```diff
@@ -30,18 +34,19 @@ Progress::~Progress() {
 void Progress::Increment(uint64_t amount, std::string update) {
   if (amount > 0) {
     std::lock_guard<std::mutex> guard(m_mutex);
+    m_details = update;
```
Good point, I'll add the `std::optional` for `Increment` and remove this parameter from `ReportProgress` 👍🏾

chelcassanova committed:
Use std::optional for detail update, remove std::optional as member
variable for total

Change the parameter for `Increment` to use a std::optional string for
the updated details as well as remove the string parameter for
`ReportProgress` as its no longer necessary since all details are being
stored in a member variable at object creation. Also reverts Progress's
`m_total` to being a uint64_t instead of having both the member variable
and construction parameter both be a std::optional<uint64_t>

chelcassanova commented on lldb/include/lldb/Core/Progress.h:
```diff
@@ -90,10 +90,11 @@ class Progress {
   /// @param [in] an optional message associated with this update.
-  void Increment(uint64_t amount = 1, std::string update = {});
+  void Increment(uint64_t amount = 1,
+                 std::optional<std::string> updated_detail = {});
```
`updated_detail` is still kept as an empty string by default, though I'm unsure if the empty string or `std::nullopt` is preferred here

clayborg commented on lldb/source/Core/Progress.cpp:
```diff
@@ -35,25 +42,29 @@ Progress::~Progress() {
   if (amount > 0) {
     std::lock_guard<std::mutex> guard(m_mutex);
+    if (updated_detail)
+      m_details = *updated_detail;
```
We could use std::move() here to avoid a copy?

clayborg commented on lldb/source/Core/Progress.cpp:
```diff
@@ -35,25 +42,29 @@ Progress::~Progress() {
-    Debugger::ReportProgress(m_id, m_title, std::move(update), m_completed,
-                             m_total, m_debugger_id);
+    // complete, and only modify m_complete is m_total isn't null.
+    if (m_total)
+      m_complete = m_completed == m_total;
```
No need for the if statement anymore, these two lines can just be:
```
m_complete = m_completed == m_total;
```

chelcassanova committed:
Use std::move when updating m_details

When `m_details` is updated with `Increment` use std::move instead of
making a copy of the updated details. Also removes the if clause that
updates m_complete in `ReportProgress`

clayborg commented on lldb/source/Core/Progress.cpp:
```diff
@@ -35,25 +42,28 @@ Progress::~Progress() {
   if (!m_complete) {
     // Make sure we only send one notification that indicates the progress is
-    // complete.
+    // complete, and only modify m_complete is m_total isn't null.
```
Is the comment out of date? revert comment?

clayborg reviewed (approved):
Thanks for the changes. On comment might be able to be reverted, but this looks good.

chelcassanova committed:
Fix comment

Fix outdated comment

medismailben reviewed (approved):
LGTM!
